### PR TITLE
Add runtime precondition check for paged attention kernel.

### PR DESCRIPTION
**Pull request by vanbasten23**
We need to verify the precondition of the paged attention kernel. The reason why this is important is that if the precondition is violated, the pallas kernel may hang or return incorrect result so then it'd be very hard to debug (is the bug in the kernel or in vLLM?)

### CODE REVIEW

**Comment by vanbasten23** [comment]
cc @bythew3i @yaochengji 
I thought about directly using https://github.com/jax-ml/jax/blob/fc5d9a4fcee2a4606f36d2d2bd517458afde24a3/jax/experimental/pallas/ops/tpu/ragged_paged_attention.py#L133 but then I have to call the function in PallasAttentionBackendImpl.forward which torch.compiled. SoI have to check before the torch compiled region.

**Comment by bvrockwell** [comment]
Hi @vanbasten23 could you help us understand what the behaviour this catches is? does it hang due to misconfiguration?

for a user trying to debug, how would they read the exceptions proposed here and adjust configs accordingly?

Thanks!

**Comment by vanbasten23** [comment]
> Hi @vanbasten23 could you help us understand what the behaviour this catches is? does it hang due to misconfiguration?
> 
> for a user trying to debug, how would they read the exceptions proposed here and adjust configs accordingly?
> 
> Thanks!

Here we validate the precondition of the ragged paged attention so it is more of a preventive check. If the precondition are violated, the kernel may misbehave: return incorrect value, or hang.

It helps the developer by setting a clear boundary: if the precondition is violated, it might be we write a bug in the vLLM code.

**Review by NickLucche — CHANGES_REQUESTED** [review]

File: vllm/v1/worker/tpu_model_runner.py [hunk 1]
```diff
@@ -485,6 +485,8 @@ def _prepare_inputs(self, scheduler_output: "SchedulerOutput"):
             self.device)
         seq_lens = self.seq_lens_cpu[:self.max_num_reqs].to(self.device)
 
+        self.validate_paged_attention_precondition(block_tables, num_reqs,
```
Comment (yaochengji) [inline vllm/v1/worker/tpu_model_runner.py:488]: I guess you put this check here because it's not very easy to check device values in the kernel. But why these tensors are xla tensors, not cpu tensors?

Comment (vanbasten23) [reply vllm/v1/worker/tpu_model_runner.py:488]: because at the end of the `_prepare_inputs`, we move the tensor from cpu to xla. Then the model runs on the device, including the pallas kernel, so the tensors are xla tensors.

I'm under the impression that checking the device value will break the torch dynamo and xla compilation.

Comment (vanbasten23) [reply vllm/v1/worker/tpu_model_runner.py:488]: > But why these tensors are xla tensors, not cpu tensors?

These tensors are cpu tensors, not xla tensors

File: vllm/v1/worker/tpu_model_runner.py [hunk 2]
```diff
@@ -507,6 +509,31 @@ def _prepare_inputs(self, scheduler_output: "SchedulerOutput"):
         logits_indices = logits_indices.to(self.device)
         return attn_metadata, logits_indices
 
+    def validate_paged_attention_precondition(self, block_tables, num_reqs,
+                                              query_start_loc, seq_lens):
```
[the remark this answers is not in the document]
Comment (vanbasten23) [reply vllm/v1/worker/tpu_model_runner.py:513]: These will be internal error. They shouldn't occur but if they do, then it's because we write a bug.

The reason why this is important is that if the precondition is violated, the pallas kernel may hang or return incorrect result so it'd be very hard to debug (is the bug in the kernel or in vLLM?)

The reason why we put it here is that the pallas.py is under torch.compile so checking the tensor value would introduce graph breaks.

File: vllm/v1/worker/tpu_model_runner.py (hunk 1 above)
Comment (yaochengji) [reply vllm/v1/worker/tpu_model_runner.py:488]: `seq_lens = self.seq_lens_cpu[:self.max_num_reqs].to(self.device)`

E.g. `seq_lens` is a xla tensor

**Review by yaochengji — CHANGES_REQUESTED** [review]

File: vllm/v1/worker/tpu_model_runner.py (hunk 1 above)
Comment (vanbasten23) [reply vllm/v1/worker/tpu_model_runner.py:484]: My bad, sorry. They are xla tensors. I moved the check to verify the cpu tensors.

Comment (bythew3i) [reply vllm/v1/worker/tpu_model_runner.py:484]: Thanks a lot @vanbasten23 for putting all these. After second thought, if it does not too much sense to add the runtime check on vllm, I think another way could be to insert checkify code before kernel inside ragged_paged_attention_v2.py but still gated by a flag (only for debug purpose).

**Comment by vanbasten23** [comment]
very strange, the change caused an error https://gist.github.com/vanbasten23/0e621f33458287f722693fe612f330fc

File: vllm/v1/worker/tpu_model_runner.py (hunk 1 above)
Comment (yaochengji) [reply vllm/v1/worker/tpu_model_runner.py:484]: @bythew3i  But the check is only related to the definition of attention for all kinds of hardware, not closely related to the attention implementation on TPU.

Comment (bythew3i) [reply vllm/v1/worker/tpu_model_runner.py:484]: @yaochengji what check are you referring to? Are these dynamic value check? I am more leaning toward adding these dyanmic check to the kernel file but calling them before pallas call.

I think we need to have debug mode in pallas.py and only run these runtime check when debug is on. Otherwise these checks slow down the runtime performance IMO.

Comment (yaochengji) [reply vllm/v1/worker/tpu_model_runner.py:484]: @bythew3i I'm referring to dynamic value check.

I agree we can add the check function in the kernel file. And the check function should work for both cpu tensor and xla tensor. We can call this check function in vLLM based on CPU tensor to make it always on.

Comment (vanbasten23) [reply vllm/v1/worker/tpu_model_runner.py:484]: Great discussion. I'd like to share my 2 cents:

1) regarding putting the dynamic check in the kernel file: Agreed that we can put to torch_xla kernel file: custom_kernels.py. One catch is that it cannot be put to the pallas kernel file because the pallas kernel file doesn't depend on torch and it only depends on jax while here all tensors are torch tensors. 

2) regarding if calling them before pallas call: that was my first attempt but then I realize pallas call is under torch compile. If we check the tensor value, it'd cause dynamo graph break and xla graph break. That's why I think it'd better to check the cpu tensors here.

Comment (NickLucche) [reply vllm/v1/worker/tpu_model_runner.py:484]: I also share @bythew3i concerns about overhead introduced by these checks. In particular I feel like most of these checks are redundant and either already checked somewhere else (or valid by construction) or so catastrophic that would have the server crash in multiple ways.
Eg `if num_reqs > max_num_reqs` means the scheduler is sending more requests than what we set. This is really bad at the core level of vllm.

I welcome these checks at kernel level or in a wrapper to help a developer understand how to use the API, but would like a way to turn them off at runtime for vllm.
